toolchains/args.bzl

```diff
@@ -0,0 +1,297 @@
+# limitations under the License.
+"""All providers for rule-based bazel toolchain config."""
+
+load(
+    "//toolchains/private:collect.bzl",
+    "collect_action_types",
+    "collect_files",
+)
+load(
+    ":toolchain_info.bzl",
+    "ActionTypeSetInfo",
+    "ArgsInfo",
+    "ArgsListInfo",
+    "FeatureConstraintInfo",
+    "ListVariableInfo",
+    "SingleVariableInfo",
+    "TYPE_INPUT_FILE",
+    "TYPE_OUTPUT_FILE",
+)
+
+visibility("public")
+
+# We have no int variables, so this shouldn't match.
+_MISSING = 3
+
+COMMON_ATTRS = {
+    "actions": attr.label_list(
+        providers = [ActionTypeSetInfo],
+        mandatory = True,
+        doc = """The actions for which this arg is used. This arg will be skipped for all actions other than the ones listed here.""",
+    ),
+    "env": attr.string_dict(
+        doc = """Environment variables to apply.""",
+    ),
+    "data": attr.label_list(
+        allow_files = True,
+        doc = "Data required for the arg. This does not include variables.",
+    ),
+    "requires": attr.label_list(
+        providers = [FeatureConstraintInfo],
+        doc = """A list of constraints, one of which must be met or an error will occur.""",
+    ),
+}
+
+def base_add_args_impl(ctx, fill, *, allowed_actions = None, allowed_actions_cause = None, storage = None):
+    """A basic implementation of args recommended to be used by all custom rules.
+
+    Args:
+        ctx: The ctx of the current rule.
+        fill: Function(self, ExecutionContext, ctx.action.Args) A function that can modify the command-line being executed.
+        allowed_actions: Optional[List[ActionTypeSetInfo]] A list of actions allowed for this action. Should be the intersection of the actions of all variables being accessed.
+        allowed_actions_cause: Optional[str] What to report of the cause of the failure if the actions didn't match.
+        storage: Any metadata to be stored, accessible via self.storage in the fill function.
+
+    Returns:
+        providers suitable for
+    """
+    actions = tuple([action.label for action in collect_action_types(ctx.attr.actions).to_list()])
+    if allowed_actions != None:
+        for action in actions:
+            if action not in allowed_actions:
+                fmt = " ".join(sorted(["  %s\n" % x for x in allowed_actions]))
+                fail("%s is not a valid action for %s, which restricts the actions to:\n%s" % (action, allowed_actions_cause, fmt))
+
+    args = ArgsInfo(
+        label = ctx.label,
+        actions = actions,
+        requires = [target[FeatureConstraintInfo] for target in ctx.attr.requires],
+        files = collect_files(ctx.attr.data),
+        env = ctx.attr.env,
+        fill = fill,
+        storage = storage,
+    )
+    return [
+        args,
+        ArgsListInfo(label = ctx.label, args = tuple([args])),
+    ]
+
+def _args_add_strings_impl(ctx):
+    def fill(self, execution_ctx, args):
+        args.add_all(self.storage)
```

Comment:
This may be more memory efficient if we added another shared `Args` instance instead of mutating a single one. We don't have to do this now and should definitely benchmark, but is this a change we could make later without affecting public API? It would be great if we could limit exposure to the point where such changes would remain possible, at least at this early stage.

Reply:
I'm a little confused because I'm not sure what you mean by "added another shared `Args` instance instead of mutating a single one". I remember you brought up a point a few months ago about potentially saving memory using a single `Args` per action, instead of one `Arg` per `args_add_*` rule. I'm guessing that's what you're referring to here?

If so, yes, this would be trivial to change without changing the API ([source](https://github.com/matts1/rules_toolchains/blob/72b23a7086263427e25fbd5eabfe97ec0bbeb61f/toolchains/generic_rule.bzl#L81-L83)).

If you meant something else, could you please elaborate further.